std-bits/table/src/main/java/org/enso/table/data/table/join/LookupJoin.java

```diff
@@ -0,0 +1,205 @@
+        List<Object> exampleValues = IntStream.range(0, keyColumnNames.size()).mapToObj(key::get).toList();
+        throw new NonUniqueLookupKey(keyColumnNames, exampleValues, groupSize);
+      }
+    }
+  }
+
+  private static void checkNullsInKey(Column[] lookupKeyColumns) {
+    for (Column c : lookupKeyColumns) {
+      boolean containsNulls = c.getStorage().countMissing() > 0;
+      if (containsNulls) {
+        Storage<?> s = c.getStorage();
+        int missingRow = -1;
+        for (int i = 0; i < s.size(); i++) {
+          if (s.isNa(i)) {
+            missingRow = i;
+            break;
+          }
+        }
+
+        assert missingRow != -1;
+        final int row = missingRow;
+        List<Object> exampleValues =
+            IntStream.range(0, lookupKeyColumns.length).mapToObj(i -> (Object) s.getItemBoxed(row)).toList();
+        throw new NullValuesInKeyColumns(exampleValues);
+      }
+    }
+  }
+
+  private Table join() {
+    List<LookupOutputColumn> outputColumns = columnDescriptions.stream().map(this::prepareOutputColumn).toList();
+    List<LookupOutputColumn.MergeColumns> columnsToMerge =
+        outputColumns.stream().filter(LookupOutputColumn.MergeColumns.class::isInstance).map(LookupOutputColumn.MergeColumns.class::cast).toList();
+
+    // We have columns to merge only if unmatched rows are expected. If unmatched rows are not allowed, all lookup
+    // columns will completely replace old values, so we can rely on the OrderMask optimization which is more efficient.
+    assert allowUnmatchedRows || columnsToMerge.isEmpty();
+
+    boolean needsOrderMask = outputColumns.stream().anyMatch(LookupOutputColumn.AddFromLookup.class::isInstance);
+    int[] orderMask = needsOrderMask ? new int[baseTableRowCount] : null;
+
+    for (int i = 0; i < baseTableRowCount; i++) {
+      // Find corresponding row in the lookup table
+      int lookupRow = findLookupRow(i);
+
+      assert allowUnmatchedRows || lookupRow != Index.NOT_FOUND;
+
+      // Merge columns replacing old values
+      for (LookupOutputColumn.MergeColumns mergeColumns : columnsToMerge) {
+        Object itemToAdd;
+        if (lookupRow != Index.NOT_FOUND) {
+          itemToAdd = mergeColumns.lookupReplacement.getItemBoxed(lookupRow);
+        } else {
+          itemToAdd = mergeColumns.original.getItemBoxed(i);
+        }
+        mergeColumns.builder.appendNoGrow(itemToAdd);
+      }
+
+      // Prepare order mask for new columns / fully-replaced columns
+      if (needsOrderMask) {
+        orderMask[i] = lookupRow;
+      }
+    }
+
+    Column[] columns = outputColumns.stream().map(c -> c.build(orderMask)).toArray(Column[]::new);
+    return new Table(columns);
+  }
+
+  private int findLookupRow(int baseRowIx) {
+    UnorderedMultiValueKey key = makeTableRowKey(baseRowIx);
+    List<Integer> lookupRowIndices = lookupIndex.get(key);
+    if (lookupRowIndices == null || lookupRowIndices.isEmpty()) {
```

Comment:
is an empty index possible?

Reply:
theoretically not, I used to not have this second condition, I guess I can remove it indeed